### PR TITLE
chore: update input validation

### PR DESCRIPTION
**Pull request by rsmayda**
Checklist:

- [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/update-user-roles.json [hunk 1]
```diff
@@ -10,22 +10,17 @@
...
-      ]
+      "enum": ["INTERNAL", "EXTERNAL"]
     }
   },
-  "required": [
-    "id",
-    "rev",
-    "userType"
-  ]
+  "required": ["id", "rev", "userType"]
```
Comment (Bingjiling) [inline addons/addon-base-raas/packages/base-raas-services/lib/schema/update-user-roles.json:25]: Not sure if 'userType' should be required?

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js [hunk 2]
```diff
@@ -701,6 +671,65 @@ describe('studyService', () => {
   });
 
   describe('update', () => {
+    it('should fail to update resource list of non-Open Data study', async () => {
```
Comment (Bingjiling) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js]: Nice move!

File: openapi.yaml [hunk 3]
```diff
@@ -3400,11 +3400,11 @@ components:
           example: 2
```
Comment (Bingjiling) [inline openapi.yaml:3401]: Other schema validation updates will need to be reflected in openapi.yaml as well. It can be done in a future PR.

**Review by Bingjiling — COMMENTED** [review]
Looks good! One minor thing on adding userType as required for the update schema. I'm not sure where the method is used and if adding userType as required would break anything.

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/update-user-roles.json (hunk 1 above)
Comment (rsmayda) [reply addons/addon-base-raas/packages/base-raas-services/lib/schema/update-user-roles.json:25]: It was there before just reformatted -- no material change

File: openapi.yaml (hunk 3 above)
Comment (rsmayda) [reply openapi.yaml:3401]: Updated

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json [hunk 4]
```diff
@@ -5,20 +5,26 @@
...
         },
         "startDate": {
-          "type": "integer"
+          "type": "integer",
+          "minimum": 0
         },
         "endDate": {
-          "type": "integer"
+          "type": "integer",
+          "minimum": 0
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json:27]: Do we need a test for `startDate` and `endDate`?

File: addons/addon-environment-sc-api/packages/environment-type-mgmt-services/lib/environment-type/__tests__/env-type-config-service.test.js [hunk 5]
```diff
@@ -215,18 +215,58 @@ describe('EnvTypeService', () => {
...
+            ParameterKey: 'someProperty',
+            ParameterType: 'String',
+          },
+        ],
+      };
+      envTypeService.mustFind.mockImplementationOnce(() => envType);
+      service.audit = jest.fn();
+
+      // OPERATE and CHECK
+      await expect(service.create({}, newConfig.id, newConfig)).rejects.toThrow(
```
Comment (jn1119) [inline addons/addon-environment-sc-api/packages/environment-type-mgmt-services/lib/environment-type/__tests__/env-type-config-service.test.js:249]: Were you able to add validation in `id` provided in request?

**Review by SanketD92 — CHANGES_REQUESTED** [review]
Unit tests seem to be failing :(

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json [hunk 6]
```diff
@@ -29,6 +35,7 @@
         },
         "notificationEmail": {
           "type": "string",
+          "maxLength": 512,
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json:38]: Add that `<hack>` test case for `notificationEmail`

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-study.json [hunk 7]
```diff
@@ -11,15 +11,18 @@
     },
     "name": {
       "type": "string",
-      "maxLength": 2048
+      "minLength": 1,
+      "maxLength": 300,
+      "pattern": "^([^<>{}]+)$"
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/schema/create-study.json:16]: Do we need tests for `name`, `description` and other string type fields?

**Review by jn1119 — COMMENTED** [review]
I didn't go through the entire PR but it seems like for unit testing we need to define the criteria. Should we add unit tests for all the `string` fields?

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json (hunk 4 above)
Comment (rsmayda) [reply addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json:27]: From a schema standpoint we can't due much more than this. We also already have it https://github.com/awslabs/service-workbench-on-aws/blob/mainline/addons/addon-base-raas/packages/base-raas-services/lib/budgets/budgets-service.js#L62

File: addons/addon-environment-sc-api/packages/environment-type-mgmt-services/lib/environment-type/__tests__/env-type-config-service.test.js (hunk 5 above)
Comment (rsmayda) [reply addons/addon-environment-sc-api/packages/environment-type-mgmt-services/lib/environment-type/__tests__/env-type-config-service.test.js:249]: Talked offline; yes the id is validated!

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json (hunk 4 above)
Comment (jn1119) [reply addons/addon-base-raas/packages/base-raas-services/lib/schema/create-budget.json:27]: [non blocking] nit: I just meant adding a unit test when startDate and endDate are negative.

File: addons/addon-base-raas/packages/base-raas-services/lib/schema/create-study.json [hunk 8]
```diff
@@ -42,10 +46,12 @@
           "additionalProperties": false,
           "properties": {
             "arn": {
-              "type": "string"
+              "type": "string",
+              "maxLength": 2048,
+              "pattern": "^(arn:aws[a-zA-Z-]*:[a-zA-Z0-9-/.:_?*]+)$"
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/schema/create-study.json:51]: nit: Add test for `arn` too?

**Review by jn1119 — APPROVED** [review]
Approved. LGTM overall. Left one minor comment

**Review by SanketD92 — APPROVED** [review]
Looks good.
